#1408 improving assignment

nmadruga opened the pull request:
Fixes #1408 and fixes #1423 - new bug introduced in v0.4.6-13 

### Briefing about extra commits

This branch was based on `build-2022-11-03A`, but doesn't seem to be very much different from `develop`.

### Changes

- Added `scrollable` in Dropdowns - UX very desired feature
- Improving to remove option "NOT ASSIGNED" from dropdown 
- Keep a constant shared for NOT_ASSIGNED -1 for when nothing is selected
- Fix how is showing the selected Assigned, something was wrong with the hook `getAssignmentOptions` that I didn't want to dive in so much, but adding the `isSubmitted` helped to show correctly the selected assignment before clicking on the Apply
- Fix problem where Re-assignment wasn't showing all available options
- Fix nor reloading after unassigned reviewer

I tried not to change much the code, but be simplifying the structure with the removal of 1 Component and combining more variables to new hook `useLoadAssignment`.

nmadruga commented on src/components/Review/ReviewSectionRowAction.tsx:
```diff
@@ -151,9 +145,14 @@ const GenerateActionButton: React.FC<ReviewSectionComponentProps> = ({
 
     try {
-      if (isFinalDecision)
+      if (isFinalDecision && previousAssignment) {
+        const remakeReview = useRemakePreviousReview({
+          reviewStructure,
+          reviewAssignment,
+          previousAssignment,
+        })
```
Moved this to here since it's only used for stage of FinalDecision...

nmadruga commented on src/containers/Review/assignment/AssignAll.tsx:
```diff
@@ -30,16 +33,25 @@ const AssignAll: React.FC<AssignAllProps> = ({ assignments, setReviewerForAll, c
-          setSelected(value as number | string)
-        }}
-      />
+        text={
+          options.find(({ value }) => value === selected)?.text || strings.ASSIGNMENT_NOT_ASSIGNED
+        }
```
The previous `value={selected}` wasn't working after changing to use the header ASSIGNMENT_NOT_ASSIGNED

nmadruga commented on src/containers/Review/assignment/AssignAll.tsx:
```diff
@@ -30,16 +33,25 @@ const AssignAll: React.FC<AssignAllProps> = ({ assignments, setReviewerForAll, c
+          {options.map((assignee) => (
+            <Dropdown.Item
+              {...assignee}
+              onClick={(_, { value }) => {
+                setReviewerForAll(value as number)
+                setSelected(value as number | string)
+              }}
```
This is the same as `onChange` was previously doing, but didn't work with this new shape of Dropdown

nmadruga commented on src/containers/Review/assignment/AssigneeDropdown.tsx:
```diff
@@ -1,30 +1,38 @@
-      text: string
-    }[]
-  }
+  assignmentOptions: AssignmentOptions
```
Just using existing type

nmadruga commented on src/containers/Review/assignment/AssigneeDropdown.tsx:
```diff
@@ -1,30 +1,38 @@
+          />
+        ))}
+      </Dropdown.Menu>
+    </Dropdown>
```
Very similar to what you see above for AssignAll, we should be able to combine this two components, but not right now

nmadruga commented on src/containers/Review/assignment/AssignmentSectionRow.tsx:
```diff
@@ -128,7 +126,7 @@ const AssignmentSectionRow: React.FC<AssignmentSectionRowProps> = ({
         </Grid.Column>
         <Grid.Column className="centered-flex-box-row" width={9}>
-          {originalAssignee ? (
+          {originalAssignee && assignmentOptions.isSubmitted ? (
```
This was a **new** problem, happens after the selection of the assigned reviewer, it was interpreting the same way as an existing Assignment on the database, so I added this flag to determine when this is submitted assignment or un-submitted

nmadruga commented on src/containers/Review/assignment/AssignmentSectionRow.tsx:
```diff
@@ -154,12 +152,11 @@ const AssignmentSectionRow: React.FC<AssignmentSectionRowProps> = ({
-            isLastLevel={isLastLevel}
-            previousAssignee={assignmentOptions.selected}
+            isLastLevel={isLastLevel()}
             assignedSections={assignedSections}
             setAssignedSections={setAssignedSections}
+            assignmentOptions={assignmentOptions}
```
There was some extra variables not needed for Re-assignment

I was investigating why it used to show only the same assignment on the dropdown, I found out the problem and ended up removing some extra props that could be already used with `assignmentOptions`

nmadruga commented on src/containers/Review/assignment/AssignmentTab.tsx:
```diff
@@ -1,82 +1,55 @@
-    applicationId: fullStructure.info.id,
-    serial: fullStructure.info.serial,
-    userId: currentUser?.userId as number,
+  const {
+    error,
+    loading,
+    assignmentsFiltered,
+    assignmentInPreviousStage,
+    assignmentGroupedLevel,
+    isFullyAssigned,
+  } = useLoadAssignments({
+    info: fullStructure.info,
+    sectionCodes,
+    filters,
   })
```
New hook, where most of the code removed from here is now

nmadruga commented on src/containers/Review/assignment/AssignmentTab.tsx:
```diff
@@ -103,21 +74,59 @@ const AssignmentTab: React.FC<{
+  const AssignmentSection: React.FC<SectionDetails> = ({ id, title, code }) => {
+    const { reviewStructuresState } = useReviewStructureState()
+    return (
+      <Segment key={id}>
+        <Header className="section-title" as="h5" content={title} />
+        {Object.entries(assignmentGroupedLevel).map(([level, assignments]) => (
+          <div className="flex-column" key={`assignment-group-level-${level}`}>
+            {fullStructure.info.outcome === ApplicationOutcome.Pending && (
+              <AssignmentSectionRow
+                assignments={assignments}
+                sectionCode={code}
+                reviewLevel={Number(level)}
+                structure={fullStructure}
+                assignedSections={assignedSectionsByLevel[level] || defaultAssignedSections}
+                setAssignedSections={(assignedSections) =>
+                  setAssignedSectionsByLevel({
+                    ...assignedSectionsByLevel,
+                    [level]: assignedSections,
+                  })
+                }
+                setEnableSubmit={setEnableSubmit}
+                setAssignmentError={setAssignmentError}
+              />
+            )}
+            {(assignments as AssignmentDetails[]).map((assignment) =>
+              reviewStructuresState[assignment.id] && assignment.isCurrentUserReviewer ? (
+                <ReviewSectionRow
+                  key={`review-row-section-${code}-assignment-${assignment.id}`}
+                  sectionId={id}
+                  fullStructure={fullStructure}
+                  reviewAssignment={assignment}
+                  previousAssignment={assignmentInPreviousStage}
+                />
+              ) : null
+            )}
+          </div>
+        ))}
+      </Segment>
+    )
+  }
```
This used to be another component `AssignmentRows` and it was confusing so I thought it would be better to move in here

CarlosNZ commented:
> * Fix how is showing the selected Assigned, something was wrong with the hook `getAssignmentOptions` that I didn't want to dive in so much, but adding the `isSubmitted` helped to show correctly the selected assignment before clicking on the Apply
> 
> * Fix problem where Re-assignment wasn't showing all available options
> 
> * Fix nor reloading after unassigned reviewer

Looks good -- would you be able to provide a snapshot where these bugs are apparent, so we can compare the difference this branch makes? Thanks.

nmadruga commented:
> > * Fix how is showing the selected Assigned, something was wrong with the hook `getAssignmentOptions` that I didn't want to dive in so much, but adding the `isSubmitted` helped to show correctly the selected assignment before clicking on the Apply
> > * Fix problem where Re-assignment wasn't showing all available options
> > * Fix nor reloading after unassigned reviewer
> 
> Looks good -- would you be able to provide a snapshot where these bugs are apparent, so we can compare the difference this branch makes? Thanks.

The later 2 problems can be checked in any snapshot running on branch prior to v0.4.6-13.

nmadruga commented:
The one about adding `isSubmitted` I think might not be required... But was required as explained [here](https://github.com/openmsupply/conforma-web-app/pull/1421#discussion_r1013535902)

"This was a new problem, happens after the selection of the assigned reviewer, it was interpreting the same way as an existing Assignment on the database, so I added this flag to determine when this is submitted assignment or un-submitted"

nmadruga commented on src/utils/hooks/useCreateFinalDecisionReview.ts:
```diff
@@ -54,9 +54,12 @@ const useRemakePreviousReview: UseRemakePreviousReview = ({
       ({ isPendingReview, thisReviewLatestResponse, response, lowerLevelReviewLatestResponse }) => {
-        const applicationResponseId = previousAssignment.level > 1 ? undefined : response?.id
+        const applicationResponseId =
+          !previousAssignment || previousAssignment.level > 1 ? undefined : response?.id
         const reviewResponseLinkId =
-          previousAssignment.level === 1 ? undefined : lowerLevelReviewLatestResponse?.id
+          !previousAssignment || previousAssignment.level === 1
+            ? undefined
+            : lowerLevelReviewLatestResponse?.id
```
Part of fix #1423

nmadruga commented on src/utils/hooks/useCreateFinalDecisionReview.ts:
```diff
@@ -12,16 +12,16 @@ import { useGetFullReviewStructureAsync } from './useGetReviewStructureForSectio
 type PromiseReturnType = ReturnType<UseRemakePreviousReviewMutationReturnType[0]>
 // hook used to restart a review, , as per type definition below (returns promise that resolve with mutation result data)
-type UseRemakePreviousReview = (props: {
+type UseCreateFinalDecisionReview = (props: {
```
Renaming the hook to be clearer what's doing

CarlosNZ replied:
So much clearer! :)

nmadruga commented on src/utils/hooks/useCreateFinalDecisionReview.ts:
```diff
@@ -12,16 +12,16 @@ import { useGetFullReviewStructureAsync } from './useGetReviewStructureForSectio
   reviewStructure: FullStructure
   reviewAssignment: AssignmentDetails
-  previousAssignment: AssignmentDetails
+  previousAssignment?: AssignmentDetails
```
Part of fix #1423

nmadruga commented on src/containers/Review/assignment/useGetAssignmentOptions.ts:
```diff
@@ -57,15 +63,22 @@ const useGetAssignmentOptions = () => {
-      value: NOT_ASSIGNED,
-      text: strings.ASSIGNMENT_NOT_ASSIGNED,
-    })
+    // Move current user "Yourself" to top of list
+    const currentUserOption = assigneeOptions.options.find(
+      (option) => option.value === currentUser?.userId
+    )
+
+    // Remove currentUser from the list and add at the top
+    const sortedOptions = assigneeOptions.options.filter(
+      (option) => option.key !== currentUser?.userId
+    )
+    if (currentUserOption) sortedOptions.unshift(currentUserOption)
+    assigneeOptions.options = sortedOptions
```
Suggestion you gave @CarlosNZ 

The missing part (to remove the duplicate user) was this:
```
const sortedOptions = assigneeOptions.options.filter(
      (option) => option.key !== currentUser?.userId
    )
```

nmadruga commented:
Ready for another review - before we can build version to install in 50004 today

CarlosNZ commented on src/containers/Review/assignment/AssignAll.tsx:
```diff
@@ -2,20 +2,23 @@ import React, { useState } from 'react'
   const options = getReviewerList(
     assignments,
-    currentUser?.userId ?? 0,
+    currentUser?.userId || 0,
```
Just interested -- why `||` rather than `??` ? (They would both behave exactly the same in this situation)

nmadruga replied:
Yes, I think that was required. I'll revert...

CarlosNZ commented on src/containers/Review/assignment/AssigneeDropdown.tsx:
```diff
@@ -1,28 +1,27 @@
       options={options}
       value={selected}
       disabled={isCompleted}
+      labeled
```
What exactly is `labeled` doing here? I don't see any difference.

nmadruga replied:
Oh, Cool! I think it was used to show the selected option before. I'll remove

CarlosNZ commented on src/containers/Review/assignment/AssignmentSectionRow.tsx:
```diff
@@ -138,7 +136,7 @@ const AssignmentSectionRow: React.FC<AssignmentSectionRowProps> = ({
             />
           ) : (
-            assignmentOptions.options.length > 1 && (
+            assignmentOptions.options.length > 0 && (
```
Not sure I understand this change. Shouldn't it still show if there's only 1 option?

nmadruga replied:
Yes, so it would show if there is at least one option (instead of at least 2 options as before - because it was considering the NOT ASSIGNED not selectable option)

CarlosNZ replied:
Ah right -- got it. 👍

CarlosNZ reviewed (approved):
Okay, looks good thanks. It'll be good to get this on the server and see if any other problems arise. But seems pretty solid I think.

Couple of very minor question -- the only one that's potentially important is the one about the options.length > 1

CarlosNZ commented:
Sorry, one more little thing. Maybe not part of this PR, but I noticed this message:
![Screenshot 2022-11-07 at 3 07 40 PM](https://user-images.githubusercontent.com/5456533/200212690-9706ed18-eb65-4332-aaed-c9b2e5a0e32b.png)

The message doesn't really make sense -- it shouldn't include the name when its "self-asssignable" -- you only "self-assign" yourself, so should just say "Can be self-assigned". 

Ignore it for now if it's a hassle, we can deal with it elsewhere.

CarlosNZ commented:
Sorry, one more little thing. Maybe not part of this PR, but I noticed this message:
![Screenshot 2022-11-07 at 3 07 40 PM](https://user-images.githubusercontent.com/5456533/200212690-9706ed18-eb65-4332-aaed-c9b2e5a0e32b.png)

The message doesn't really make sense -- it shouldn't include the name when its "self-asssignable" -- you only "self-assign" yourself, so should just say "Can be self-assigned". 

Ignore it for now if it's a hassle, we can deal with it elsewhere.

nmadruga commented:
Ok - Carl I have replied to your changes suggested now. I have also added the `AssignAll` similar changes and added sorting as well - I think that seems like all #1408 would be completed by this PR now. I'll edit the description.


The only sad part is that Re-assignment isn't showing on the same order... Even though it uses the same component (As assigning) and the same options... So very odd. But I thinks it's not worth investigating more than the 5 minutes I have.
